### ISSUE TITLE
Question about finetuning SAM2 using own image dataset

### ISSUE THREAD

**Issue by hpichlerbio** [facebookresearch/sam2#347]
I´ve been really looking forward to getting an official implementation for finetuning SAM, and I'm really glad to finally see this. 
However, I don´t fully understand the tutorial provided in the README for the training scripts. I want to finetune the model on my own dataset, which only contains images, not videos. The last chapter in the README tells me, that it should be possible, but you only provide information on how to finetune it using images and videos simultaneously and also specifically just for the SA-1B and SA-V dataset, which I'm not interested in. 

Can you provide some more information on if using my own dataset is possible here, and if yes, how can I do that. For me, it wouldn´t be an issue if I would need to just convert my dataset into the SA-1B format, but that would be nice to know. Could you maybe provide a simple config example, which provides the necessary information to start a training on just an image dataset?

Thank you in advance!

**Comment by haithamkhedr** [comment]
Hi @hpichlerbio, thanks for your interest. One way to finetune on your custom **image** only dataset is to follow the last [section](https://github.com/facebookresearch/sam2/tree/main/training#training-on-images-and-videos) in the training README but remove the video dataset from the mix. The config would look something like this (removing the video dataset from the mix):

```yaml
data:
  train:
    _target_: training.dataset.sam2_datasets.TorchTrainMixedDataset 
    phases_per_epoch: ${phases_per_epoch} # Chunks a single epoch into smaller phases
    batch_sizes: # List of batch sizes corresponding to each dataset
    - ${bs1} # Batch size of dataset 1
    datasets:
    # Custom Image dataset
    - _target_: training.dataset.vos_dataset.VOSDataset
      training: true
      video_dataset:
        _target_: training.dataset.vos_raw_dataset.CustomImageDataset # Your custom Dataset class
        img_folder: ${path_to_img_folder}
        gt_folder: ${path_to_gt_folder}
        file_list_txt: ${path_to_train_filelist} # Optional
      sampler:
        _target_: training.dataset.vos_sampler.RandomUniformSampler
        num_frames: 1
        max_num_objects: ${max_num_objects_per_image}
      transforms: ${image_transforms}
    shuffle: True
    num_workers: ${num_train_workers}
    pin_memory: True
    drop_last: True
    collate_fn:
    _target_: training.utils.data_utils.collate_fn
    _partial_: true
    dict_key: all
```

Note that if you'd like to use your custom dataset, you should implement your own dataset class (similar to [`SA1BRawDataset`](https://github.com/facebookresearch/sam2/blob/52198ead0eb13ae8270bea6ca768ef175f5bf167/training/dataset/vos_raw_dataset.py#L148)). If your dataset is in SA1B format, you can directly use `SA1BRawDataset`. Please let me know if you have further questions

**Comment by hpichlerbio** [comment]
Thank you, this helps me a lot!

**Comment by deyang2000** [comment]
Thank you for your response. Based on that, I would like to ask: how should the `image_transforms` be constructed?
original code：transforms: ${vos.train_transforms}
```
#` Video transforms
vos:
  train_transforms:
    - _target_: training.dataset.transforms.ComposeAPI
      transforms:
        - _target_: training.dataset.transforms.RandomHorizontalFlip
          consistent_transform: True
        - _target_: training.dataset.transforms.RandomAffine
          degrees: 25
          shear: 20
          image_interpolation: bilinear
          consistent_transform: True
        - _target_: training.dataset.transforms.RandomResizeAPI
          sizes: ${scratch.resolution}
          square: true
          consistent_transform: True
        - _target_: training.dataset.transforms.ColorJitter
          consistent_transform: True
          brightness: 0.1
          contrast: 0.03
          saturation: 0.03
          hue: null
        - _target_: training.dataset.transforms.RandomGrayscale
          p: 0.05
          consistent_transform: True
        - _target_: training.dataset.transforms.ColorJitter
          consistent_transform: False
          brightness: 0.1
          contrast: 0.05
          saturation: 0.05
          hue: null
        - _target_: training.dataset.transforms.ToTensorAPI
        - _target_: training.dataset.transforms.NormalizeAPI
          mean: [0.485, 0.456, 0.406]
          std: [0.229, 0.224, 0.225]
```
A version from GPT is:

```
image_transforms:
  train_transforms:
    - _target_: dataset.transforms.ComposeAPI
      transforms:
        - _target_: dataset.transforms.RandomHorizontalFlip
          p: 0.5  
        - _target_: dataset.transforms.RandomAffine
          degrees: 25
          shear: (5, 5)  
          image_interpolation: bilinear
        - _target_: dataset.transforms.RandomResizedCrop
          size: ${image_size}
          scale: (0.08, 1.0)  
          ratio: (3/4, 4/3)
          image_interpolation: bilinear
        - _target_: dataset.transforms.ColorJitter
          brightness: 0.1
          contrast: 0.03
          saturation: 0.03
          hue: 0.01
        - _target_: dataset.transforms.RandomGrayscale
          p: 0.1 
        - _target_: dataset.transforms.ToTensorAPI
        - _target_: dataset.transforms.NormalizeAPI
          mean: [0.485, 0.456, 0.406]
          std: [0.229, 0.224, 0.225]
```
Is the method mentioned above correct? 
Looking forward to your response.

**Comment by hpichlerbio** [comment]
Hey @deyang2000, what do you mean exactly? For me it worked to just put in `${vos.train_transforms}` at the `data.train.dataset.transforms` key. This references to the original transforms in the config file. I just kept them the way they were. 

Also, if you are trying this on your own image set, I would recommend you to just transform your data into the SA1B data. This worked well for me.

**Comment by deyang2000** [comment]
I think your approach is feasible. GPT's solution is basically the same as the original code, except that it removed the "consistent_transform" function.

**Comment by samra-irshad** [comment]
Hi, i am trying to fine tune sam2 for my own dataset. Initially, i have tried writing my own class but i have to change many transform functions because they have been written for video data. Now i want to organize my data the way SA1B data is organized so that i can use `SA1BRawDataset` class. I want to know how can i organize my data in format of SA1B data? SA1B data is very large so i am unable to download it and i am not sure how it is structured. For my data, I have images and binary segmentations.

**Comment by deyang2000** [comment]
> Hi, i am trying to fine tune sam2 for my own dataset. Initially, i have tried writing my own class but i have to change many transform functions because they have been written for video data. Now i want to organize my data the way SA1B data is organized so that i can use `SA1BRawDataset` class. I want to know how can i organize my data in format of SA1B data? SA1B data is very large so i am unable to download it and i am not sure how it is structured. For my data, I have images and binary segmentations.



You can download part of the dataset at this link: [https://opendatalab.com/OpenDataLab/SA-1B/tree/main/raw](url).
Each file is around 10 GB.
Before you unzip the downloaded folder, change the file.tar to file.tar.gz, then you can directly unzip a single file. 
It should look like this:
![image](https://github.com/user-attachments/assets/697787e1-bf4b-47ae-8104-1d2a56cfc933)

**Comment by BENgoooo** [comment]
> Hey @deyang2000, what do you mean exactly? For me it worked to just put in `${vos.train_transforms}` at the `data.train.dataset.transforms` key. This references to the original transforms in the config file. I just kept them the way they were.
> 
> Also, if you are trying this on your own image set, I would recommend you to just transform your data into the SA1B data. This worked well for me.

Hey，I am highly indebted to you for posing this question, cause i have the same problem. But now i have some problem with the exchange from Labelme to SA-1B, I got method with gpt, but I'm not sure if that's right. do you have some method? What is the format of your own data set? I'd appreciate it if you could give me some advice

**Comment by hpichlerbio** [comment]
@BENgoooo this looks right to me. For me personally, this looks a bit too complicated. My annotations look all like the following:

{
    "image": {
        "image_id": 1,
        "width": 1400,
        "height": 1400,
        "file_name": "1.jpg"
    },
    "annotations": [
        {
            "area": 45499,
            "segmentation": {
                "size": [
                    1400,
                    1400
                ],
                "counts": "{Counts as byte string}"
            }
        }
    ]
}

This works for training the model. My corresponding file structure is this:


dataset

- train
    - images
        - 1.jpg
        - ...
        - x.jpg
     - labels
        - 1.json
        - ...
        - x.json
- val
- test


Hope this helps

**Comment by GeorgiaA** [comment]
Hello @hpichlerbio, @BENgoooo @deyang2000 @samra-irshad. Thank you for this useful discussion. I need to determine how much GPU time I need on the cloud to fine-tune SAM 2 (so I can provide an estimated cost to my explorer) and I'm pretty unsure at the moment. Could any of you provide some information on how much time it took for you to fine-tune the model. I'm interested in information such as:

1. What hardware did you use (e.g. did you use an NVIDIA A100 GPU as it suggests in the training resource)?
2. How much data did you have?
3. How many epochs did you need to train for?
4. How long did it take to fine-tune the model in total (or per epoch)?

Any information would greatly be appreciated so I can try and figure out how long it will take me.

Thanks

**Comment by deyang2000** [comment]
So far, I haven't officially started the fine-tuning process. 
Previously, I conducted a trial on a 4090 GPU（24G） using part of the MOSE dataset, modifying the configuration to set num_frames: 4 (originally num_frames: 8), and found that it could train normally. 
Additionally, I repeated the same experiment on a machine with an A6000 GPU（48G）, with num_frames set to 4 (originally 8), and it also trained normally. What puzzles me is that despite the A6000 having twice the memory size of the 4090, the same configuration leads to both GPUs exhausting their memory. I haven't yet conducted a detailed comparison of their inference speed and performance.

**Comment by atutuncuogluu** [comment]
> @BENgoooo this looks right to me. For me personally, this looks a bit too complicated. My annotations look all like the following:
> 
> { "image": { "image_id": 1, "width": 1400, "height": 1400, "file_name": "1.jpg" }, "annotations": [ { "area": 45499, "segmentation": { "size": [ 1400, 1400 ], "counts": "{Counts as byte string}" } } ] }
> 
> This works for training the model. My corresponding file structure is this:
> 
> dataset
> 
> * train
>   
>   * images
>     
>     * 1.jpg
>     * ...
>     * x.jpg
>   * labels
>     
>     * 1.json
>     * ...
>     * x.json
> * val
> * test
> 
> Hope this helps

Could you explain what to do after converting the dataset to the SA1B format? That part is a bit confusing for me.

Also, after downloading and setting up SAM2 as a repository, I found the command python training/train.py -c config.yaml somewhat unclear. For example, I’m not sure how to select the SAM checkpoint (tiny, base, large) for training, or how to complete the $() parameters inside the .yaml file I prepared.

**Comment by xings19** [comment]
> Hi @hpichlerbio, thanks for your interest. One way to finetune on your custom **image** only dataset is to follow the last [section](https://github.com/facebookresearch/sam2/tree/main/training#training-on-images-and-videos) in the training README but remove the video dataset from the mix. The config would look something like this (removing the video dataset from the mix):
> 
> ```yaml
> data:
>   train:
>     _target_: training.dataset.sam2_datasets.TorchTrainMixedDataset 
>     phases_per_epoch: ${phases_per_epoch} # Chunks a single epoch into smaller phases
>     batch_sizes: # List of batch sizes corresponding to each dataset
>     - ${bs1} # Batch size of dataset 1
>     datasets:
>     # Custom Image dataset
>     - _target_: training.dataset.vos_dataset.VOSDataset
>       training: true
>       video_dataset:
>         _target_: training.dataset.vos_raw_dataset.CustomImageDataset # Your custom Dataset class
>         img_folder: ${path_to_img_folder}
>         gt_folder: ${path_to_gt_folder}
>         file_list_txt: ${path_to_train_filelist} # Optional
>       sampler:
>         _target_: training.dataset.vos_sampler.RandomUniformSampler
>         num_frames: 1
>         max_num_objects: ${max_num_objects_per_image}
>       transforms: ${image_transforms}
>     shuffle: True
>     num_workers: ${num_train_workers}
>     pin_memory: True
>     drop_last: True
>     collate_fn:
>     _target_: training.utils.data_utils.collate_fn
>     _partial_: true
>     dict_key: all
> ```
> 
> Note that if you'd like to use your custom dataset, you should implement your own dataset class (similar to [`SA1BRawDataset`](https://github.com/facebookresearch/sam2/blob/52198ead0eb13ae8270bea6ca768ef175f5bf167/training/dataset/vos_raw_dataset.py#L148)). If your dataset is in SA1B format, you can directly use `SA1BRawDataset`. Please let me know if you have further questions

I write like this, the program will output an error, i think you miss the multiplier, after i fix the bug, i run the program on SA-1B, the program will always out a warnning, like this:
```text
INFO 2024-11-02 19:49:22,049 train_utils.py: 271: Train Epoch: [0][ 100/2796] | Batch Time: 0.91 (1.31) | Data Time: 0.00 (0.44) | Mem (GB): 52.00 (52.77/54.00) | Time Elapsed: 00d 00h 02m | Losses/train_all_loss: 5.03e-01 (7.52e-01)
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
WARNING:root:Skip RandomAffine for zero-area mask in first frame after 1 tentatives
```
What should I do to reduce this warnning？
My hyperparameters:
```yaml
scratch:
  resolution: 1024
  train_batch_size: 4
  num_train_workers: 10
  num_frames: 1
  max_num_objects: 5
  base_lr: 5.0e-6
  vision_lr: 3.0e-06
  phases_per_epoch: 1
  num_epochs: 40

dataset:
  # PATHS to Dataset
  img_folder: ./SA-1B
  gt_folder: ./SA-1B 
  multiplier: 2
```
My dataset config:
```yaml
data:
    train:
      _target_: training.dataset.sam2_datasets.TorchTrainMixedDataset 
      phases_per_epoch: ${scratch.phases_per_epoch} # Chunks a single epoch into smaller phases
      batch_sizes: # List of batch sizes corresponding to each dataset
        - ${scratch.train_batch_size}
      datasets:
      # Custom Image dataset
      - _target_: training.dataset.vos_dataset.VOSDataset
        training: true
        video_dataset:
          _target_: training.dataset.vos_raw_dataset.SA1BRawDataset # Your custom Dataset class
          img_folder: ${dataset.img_folder}
          gt_folder: ${dataset.gt_folder}
        sampler:
          _target_: training.dataset.vos_sampler.RandomUniformSampler
          num_frames: ${scratch.num_frames}
          max_num_objects: 5
        transforms: ${vos.train_transforms}
        multiplier: ${dataset.multiplier}
      shuffle: True
      num_workers: ${scratch.num_train_workers}
      pin_memory: True
      drop_last: True
      collate_fn:
        _target_: training.utils.data_utils.collate_fn
        _partial_: true
        dict_key: all
```

**Comment by hpichlerbio** [comment]
> > @BENgoooo this looks right to me. For me personally, this looks a bit too complicated. My annotations look all like the following:
> > { "image": { "image_id": 1, "width": 1400, "height": 1400, "file_name": "1.jpg" }, "annotations": [ { "area": 45499, "segmentation": { "size": [ 1400, 1400 ], "counts": "{Counts as byte string}" } } ] }
> > This works for training the model. My corresponding file structure is this:
> > dataset
> > 
> > * train
> >   
> >   * images
> >     
> >     * 1.jpg
> >     * ...
> >     * x.jpg
> >   * labels
> >     
> >     * 1.json
> >     * ...
> >     * x.json
> > * val
> > * test
> > 
> > Hope this helps
> 
> Could you explain what to do after converting the dataset to the SA1B format? That part is a bit confusing for me.
> 
> Also, after downloading and setting up SAM2 as a repository, I found the command python training/train.py -c config.yaml somewhat unclear. For example, I’m not sure how to select the SAM checkpoint (tiny, base, large) for training, or how to complete the $() parameters inside the .yaml file I prepared.

Yeah this part might be a bit confusing. Inside the .yaml file you need to replace all the parameters that have an $ in the beginning (Expect the ones that start with "scratch.", these take their values from the beginning of the config, where they can be changed). For example, for the parameters data.train.datasets.dataset.datasets.video_dataset.img_folder/ gt_folder you need to put in the the path to the image folder and the the labels folder respectively (As shown above). For the checkpoint you don´t necessarily need to change anything, you can just execute the "download_ckpts.sh" script in the sam2/checkpoints/ folder. That will download the checkpoints, which are correctly referenced in the config.yaml files. Then you can start a training from the sam2 directory, e.g. using the command: `python3 training/train.py -c configs/sam2.1_training/sam2.1_hiera_b+_MOSE_finetune.yaml --use-cluster 0 --num-gpus 1`. One tip from my side: Don´t bother with the max_num_objects parameter. 3 is a good value here and increasing it does not change the performance in any way, it only increases memory usage.

I hope this helps.

**Comment by aditya-nisal** [comment]
@hpichlerbio Firstly, thank you for spurring the discussion as I can see there are so many with this issue. I was also planning on fine-tuning my custom image only dataset on sam2. I was able to run the finetuning on my dataset. However, I plan to use on edge. SO I wanted to finetune on the tiny model. Do you know what changes need to be made for switching between models?
